numpy/lib/function_base.py

```diff
@@ -514,7 +514,10 @@ def average(a, axis=None, weights=None, returned=False):
         avg = a.mean(axis)
         scl = avg.dtype.type(a.size/avg.size)
     else:
-        a = a + 0.0
+        # Cast bool, unsigned int, and int to float64 by default
```

Comment:
Don't use a multiplication for this.

 I think @jaimefrio was suggesting that you select the dtype via `dt = np.result_dtype(a, 0.0)` and then later on in the function, specify that dtype in the ufunc call.  So down below it has something like `avg = np.multiply(a, wtg).sum(axis)/scl` the suggestion is to change this to `avg = np.multiply(a, wtg, dtype=dt).sum(axis)/scl` and, of course, whatever other changes are needed to ensure the dtype handling is still or becomes correct.

Doing it this way means that you don't ever have to have a full copy of `a` around since the ufunc machinery will convert `a` for you using a reasonably sized buffer.